### PR TITLE
Features/30 convolve

### PR DESCRIPTION
**Pull request by lucaspataro**
## Description

This pull request implements a 1D convolution.

Issue/s resolved: #30 

## Changes proposed:
- Implement 1D convolution

## Type of change
- New feature (non-breaking change which adds functionality)

## Due Diligence

- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

**Comment by lucaspataro** [comment]
@krajsek @Markus-Goetz @rainman110 

I build upon @krajsek first implementation for the 1D convolution and added more use-cases (e.g. even kernel sizes)

Before i continue with the 2D convolution i would appreciate some feedback / tipps on my implementation.

**Comment by coquelin77** [comment]
If you do this the precommit wont fail:

The extra dev dependency pulls in additional tools to support the enforcement of coding conventions (Black) and to support a pre-commit hook to do the same. In order to fully use this framework, please also install the pre-commit hook with

    $ pre-commit install

In order to check compliance of your changes before even trying to commit anything, you can run

    $ pre-commit run --all-files

**Review by mtar — CHANGES_REQUESTED** [review]
@ClaudiaComito Thanks for rising this PR from the grave. The two errors I encountered are problems with the halo functions rather than convolve, I think. Speaking of halo functions, their type hints and docstrings are flawed, too.

File: heat/core/dndarray.py [hunk 1]
```diff
@@ -381,30 +377,43 @@ def get_halo(self, halo_size) -> torch.Tensor:
...
             size = self.comm.size
+
+            first_rank = 0
             next_rank = rank + 1
             prev_rank = rank - 1
             last_rank = size - 1
 
-            # if local shape is zero and it's the last process
+            if not self.balanced:
+                populated_ranks = torch.nonzero(lshape_map[:, 0]).squeeze().tolist()
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/dndarray.py]: fixed

File: heat/core/dndarray.py [hunk 2]
```diff
@@ -418,7 +427,7 @@ def get_halo(self, halo_size) -> torch.Tensor:
                 )
                 req_list.append(self.comm.Irecv(res_prev, source=next_rank))
 
-            if rank != 0:
+            if rank != first_rank:
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/dndarray.py:418]: yes that's in the unbalanced get_halo test

File: heat/core/signal.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+          convolution product is only given for points where the signals
+          overlap completely. Values outside the signal boundary have no
+          effect.
+
+    Notes
+    -----
+        TODO: fix this note and underlying API inconsistencies
+        There is  differences to the numpy convolve function:
+        The inputs are not swapped if v is larger than a. The reason is that v needs to be
+        non-splitted. This should not influence performance.
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/signal.py]: Now fixed :)

File: heat/core/signal.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    Only return the middle values of the convolution.
+    Contains boundary effects, where zeros are taken
+    into account:
+    >>> ht.convolve(a, v, mode='same')
+    DNDarray([1., 3., 3., 3., 3.])
+
+    Compute only positions where signal and filter weights
+    completely overlap:
+    >>> ht.convolve(a, v, mode='valid')
+    DNDarray([3., 3., 3.])
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/signal.py]: done

File: heat/core/signal.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    if len(a.shape) != 1 or len(v.shape) != 1:
+        raise ValueError("Only 1-dimensional input DNDarrays are allowed")
+    if a.shape[0] <= v.shape[0]:
+        raise ValueError("Filter size must not be larger than signal size")
+    if a.dtype is not v.dtype:
+        raise TypeError(
+            "Signal and filter weight must be of same dtype, are {} and {}".format(a.dtype, v.dtype)
+        )
+    if mode == "same" and v.shape[0] % 2 == 0:
+        raise ValueError("Mode 'same' cannot be use with even-sized kernel")
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/signal.py]: done

File: heat/core/signal.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    DNDarray([3., 3., 3.])
+    """
+    if not isinstance(a, DNDarray) or not isinstance(v, DNDarray):
+        raise TypeError("Signal and filter weight must be of type DNDarray")
+    if v.split is not None:
+        raise TypeError("Distributed filter weights are not supported")
+    if len(a.shape) != 1 or len(v.shape) != 1:
+        raise ValueError("Only 1-dimensional input DNDarrays are allowed")
+    if a.shape[0] <= v.shape[0]:
+        raise ValueError("Filter size must not be larger than signal size")
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/signal.py]: done

File: heat/core/tests/test_signal.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+class TestSignal(TestCase):
+    @classmethod
+    def setUpClass(cls):
+        super(TestSignal, cls).setUpClass()
+
+    def test_convolve(self):
+        full_odd = ht.array(
+            [0, 1, 3, 6, 9, 12, 15, 18, 21, 24, 27, 30, 33, 36, 39, 42, 29, 15]
+        ).astype(ht.int)
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/tests/test_signal.py:18]: no idea but it's common coming from numpy users

File: heat/core/tests/test_dndarray.py [hunk 8]
```diff
@@ -167,6 +161,42 @@ def test_gethalo(self):
                 self.assertTrue(data.halo_next is None)
                 self.assertEqual(data_with_halos.shape, (12, 0))
 
+        # test halo of imbalanced dndarray
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/tests/test_dndarray.py:165]: you're right, done

File: heat/core/signal.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+    elif mode == "valid":
+        pad_size = 0
+        gshape = a.shape[0] - v.shape[0] + 1
+    else:
+        raise ValueError("Only {'full', 'valid', 'same'} are allowed for mode")
+
+    a = pad(a, pad_size, "constant", 0)
+
+    # fetch halos and store them in a.halo_next/a.halo_prev
+    a.get_halo(halo_size)
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/signal.py:107]: I hope this is fixed now

File: heat/core/signal.py [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+
+    # apply halos to local array
+    signal = a.array_with_halos
```
[the remark this answers is not in the document]
Comment (ClaudiaComito) [reply heat/core/signal.py:110]: fixed

**Comment by ClaudiaComito** [comment]
> @ClaudiaComito Thanks for rising this PR from the grave. The two errors I encountered are problems with the halo functions rather than convolve, I think. Speaking of halo functions, their type hints and docstrings are flawed, too.

Thanks @mtar for looking into this and thinking of the edge cases. I've overhauled the docs of the halo functions, and generalized `get_halo`. Please include the code to reproduce the error messages, I'll expand the tests accordingly.

**Comment by ClaudiaComito** [comment]
@mtar I've expanded the tests. Not sure why codecov/patch complains about lines 66-69 and 71-74  in `signal.py`, they are tested and the tests run through.

**Comment by ClaudiaComito** [comment]
> @mtar I've expanded the tests. Not sure why codecov/patch complains about lines 66-69 and 71-74 in `signal.py`, they are covered and the tests run through.

Alright, I found and solved the problem, coverage is good now.

File: heat/core/signal.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    Note how the convolution operator flips the second array
+    before "sliding" the two across one another:
+    >>> a = ht.ones(10)
+    >>> v = ht.arange(3).astype(ht.float)
+    >>> ht.convolve(a, v, mode='full')
+    DNDarray([0., 1., 3., 3., 3., 3., 2.])
+    >>> ht.convolve(a, v, mode='same')
+    DNDarray([1., 3., 3., 3., 3.])
+    >>> ht.convolve(a, v, mode='valid')
+    DNDarray([3., 3., 3.])
```
Comment (mtar) [inline heat/core/signal.py:63]:
```suggestion
    Note how the convolution operator flips the second array
    before "sliding" the two across one another:
    
    >>> a = ht.ones(10)
    >>> v = ht.arange(3).astype(ht.float)
    >>> ht.convolve(a, v, mode='full')
    DNDarray([0., 1., 3., 3., 3., 3., 2.])
    >>> ht.convolve(a, v, mode='same')
    DNDarray([1., 3., 3., 3., 3.])
    >>> ht.convolve(a, v, mode='valid')
    DNDarray([3., 3., 3.])
```

Comment (ClaudiaComito) [reply heat/core/signal.py:63]: I thought we were not supposed to have blank lines within the examples section. By all means go ahead and make the change, thanks!

Comment (mtar) [reply heat/core/signal.py:63]: The docstring is rendered poorly when there is no blank line. Alternatively, you can use `\n` instead.

File: heat/core/tests/test_manipulations.py [hunk 12]
```diff
@@ -3303,12 +3303,12 @@ def test_tile(self):
         self.assertTrue(tiled_along_split.dtype is x.dtype)
 
         # test distributed tile along non-zero split axis
-        # len(reps) > x.ndim
+        len(reps) > x.ndim
```
Comment (mtar) [inline heat/core/tests/test_manipulations.py:3306]:
```suggestion
        # len(reps) > x.ndim
```

Comment (ClaudiaComito) [reply heat/core/tests/test_manipulations.py]: done

**Review by mtar — APPROVED** [review]
